src/core/server/saved_objects/service/lib/integration_tests/repository_with_proxy.test.ts

```diff
@@ -0,0 +1,463 @@
+
+    // Setup custom hapi hapiServer with h2o2 plugin for proxying
+    hapiServer = Hapi.server({
+      port: proxyPort,
+    });
+
+    await hapiServer.register(h2o2);
+    // register specific routes to modify the response and a catch-all to relay the request/response as-is
+
+    declareGetRoute(hapiServer, esHostname, esPort);
+    declareDeleteRoute(hapiServer, esHostname, esPort);
+    declarePostUpdateRoute(hapiServer, esHostname, esPort);
+
+    declareGetSearchRoute(hapiServer, esHostname, esPort);
+    declarePostSearchRoute(hapiServer, esHostname, esPort);
+    declarePostBulkRoute(hapiServer, esHostname, esPort);
+    declarePostMgetRoute(hapiServer, esHostname, esPort);
+    declarePostPitRoute(hapiServer, esHostname, esPort);
+    declarePostUpdateByQueryRoute(hapiServer, esHostname, esPort);
+
+    declarePassthroughRoute(hapiServer, esHostname, esPort);
+
+    await hapiServer.start();
+
+    // Setup kibana configured to use proxy as ES backend
+    root = kbnTestServer.createRootWithCorePlugins({
+      elasticsearch: {
+        hosts: [`http://${esHostname}:${proxyPort}`],
+      },
+      migrations: {
+        skip: false,
+      },
+    });
+    await root.preboot();
+    const setup = await root.setup();
+    registerSOTypes(setup);
+
+    start = await root.start();
+  });
+
+  afterAll(async () => {
+    await root.shutdown();
+    await hapiServer.stop({ timeout: 1000 });
+    await esServer.stop();
+  });
+
+  describe('requests when a proxy relays request/responses with the correct product header', () => {
+    let repository: ISavedObjectsRepository;
+    let myOtherType: SavedObject;
+    const myOtherTypeDocs: SavedObject[] = [];
+
+    beforeAll(async () => {
+      repository = start.savedObjects.createInternalRepository();
+
+      myOtherType = await repository.create(
+        'my_other_type',
+        { title: 'my_other_type1' },
+        { overwrite: false, references: [] }
+      );
+      for (let i = 1; i < 11; i++) {
+        myOtherTypeDocs.push({
+          type: 'my_other_type',
+          id: `myOtherTypeId${i}`,
+          attributes: { title: `MyOtherTypeTitle${i}` },
+          references: [],
+        });
+      }
+      await repository.bulkCreate(myOtherTypeDocs, {
+        overwrite: true,
+        namespace: 'default',
+      });
+    });
+
+    beforeEach(() => {
+      setProxyInterrupt(null);
+    });
+
+    it('does not alter a Not Found response if the document does not exist and the proxy returns the correct product header', async () => {
+      let customErr: any;
+      try {
+        await repository.get('my_other_type', '123');
+      } catch (err) {
+        customErr = err;
+      }
+      expect(customErr?.output?.statusCode).toBe(404);
+      expect(customErr?.output?.payload?.message).toBe(
+        'Saved object [my_other_type/123] not found'
+      );
+    });
+
+    it('returns a document if it exists and if the proxy passes through the product header', async () => {
```

Comment:
Do we really need to test a happy path? It's not related to the `unsupported product` check.

Reply:
I'd prefer to keep the happy path tests for a few reasons:
- We don't have any other integration tests for the service, except for `incrementCounter`. These tests at least increase our coverage here. They will help catch any other situations that could result in data loss when consumers rely on a 404 response meaning that something isn't found.
- The tests only need to be around as long as our workaround is. As soon as the main cause is addressed and fixed, we'll be ripping it.